### DISCUSSION TITLE
Migrate off Express

### DISCUSSION THREAD

**Discussion by kentcdodds** [epicweb-dev/epic-stack#511] [Ideas]
UPDATE: Remix is actively working on solving "the middlware problem" and when they finish that we'll be able to migrate from Express to that. So I think express is good enough to get us to that point and moving to something else in the mean time is not worth it.

Express has served the node.js community for years. I'm incredibly grateful to Express for that.

The world has moved on and Express hasn't kept up. For this reason, I want to move off Express. The Epic Stack started with express for three reasons:
1. There's a ton of middleware that makes express easy to use for common use cases
2. We don't really have much non-remix server code anyway
3. I've never felt I needed to learn or use anything else

But I think it's a good idea to upgrade to something more modern. The biggest issue is we have several things we're using in Express that we won't want to lose with whatever we move to. I'm interested in Fastify (because it's backed by great devs, it's popular with lots of plugins, and it's really fast) and Hono.js (because it's standards based).

Unfortunately I'm just not motivated enough to make this change for me to actually spend time on it myself, but if anyone else is interested, I would really like to compare the options. The biggest requirement is that we don't lose features. Pretty much every middleware in https://github.com/epicweb-dev/epic-stack/blob/e41809cd7b566ac4e96f1d21c7237528026b4ede/server/index.ts should be represented in some way by what we migrate to.

Building [an example](https://github.com/epicweb-dev/epic-stack/blob/main/docs/examples.md) is the best way to contribute to this effort because even if we decide to go a different direction, the example can still be helpful to others.

**Comment by kiliman** [comment]
Since Epic Stack will continue to use the Node.js runtime, I recommend you go with Fastify. It is the most compatible with existing Express/Connect middleware.

Hono would be a good option if Epic Stack were platform agnostic, but you lose out on the middleware ecosystem since the API is different.

**Reply by sergiodxa** [reply]
Hono has middlewares too, although not as many as Fastify and Express for sure.

It's also really fast and can be configured to be faster if you don't use Hono router and only use it to apply middlewares, basically becomes a middleware layer.

**Reply by kentcdodds** [reply]
I'm fine with Hono so long as we can do everything. Using hono actually helps with the "adaptability" guiding principle as well because it would make it easier for someone to ship to a non-node.js environment. I'm not going to bend over backward to support non-node.js stuff to be clear, but it's definitely a point for Hono if we can get what we need *and* it's more adaptable.

**Reply by trompx** [reply]
Being able to deploy on CF workers is huge. +1 for hono.

**Reply by osseonews** [reply]
I second the suggestion to deploy to Cloudflare Pages. If you do that, I think things are greatly simplified and you can plug into Cloudflare KV and D1 (soon out of beta). D1 is a Sqlite database.

**Reply by kentcdodds** [reply]
The best way to make this happen is to make an example :)

**Reply by osseonews** [reply]
Agree! Working on it. Just started on Remix and looking to migrate our entire NextJs app over. Really got convinced by your article. Just hoping (praying) that when Remix eventually implements RSC they don't make an entire architecture rewrite like Next did.

**Reply by kentcdodds** [reply]
Based on what the [Remix RFC](https://github.com/remix-run/remix/discussions/8048) says I think it's going to be much more incremental :)

**Comment by sergiodxa** [comment]
What about going with `node:http` instead? Remix will eventually have middleware by itself, in that case you won't need middlewares in the HTTP server anymore, you could just use Remix's middlewares, and you can still give access to an HTTP server by using `node:http` directly for any scenario where you need something from there.

**Reply by kentcdodds** [reply]
I would actually love that if we can accomplish everything we need without it being really complicated. Moving more stuff to Remix would be a real win.

**Reply by swalker326** [reply]
The middleware story is being discussed right now. I'm not knowledgeable enough to contribute but if there was specific things we wanted we should probably chime in. 

https://github.com/remix-run/remix/discussions/7642

**Reply by kentcdodds** [reply]
Yeah, I've looked at that proposal and it'll handle everything I want. Once that lands then almost everything will go directly into Remix anyway.
